core/server/master/src/main/java/alluxio/master/file/LostFileDetector.java

```diff
@@ -37,34 +41,44 @@ final class LostFileDetector implements HeartbeatExecutor {
   private static final Logger LOG = LoggerFactory.getLogger(LostFileDetector.class);
   private final FileSystemMaster mFileSystemMaster;
   private final InodeTree mInodeTree;
+  private final BlockMaster mBlockMaster;
 
   /**
    * Constructs a new {@link LostFileDetector}.
    */
-  public LostFileDetector(FileSystemMaster fileSystemMaster, InodeTree inodeTree) {
+  public LostFileDetector(FileSystemMaster fileSystemMaster, BlockMaster blockMaster,
+      InodeTree inodeTree) {
     mFileSystemMaster = fileSystemMaster;
     mInodeTree = inodeTree;
+    mBlockMaster = blockMaster;
     MetricsSystem.registerCachedGaugeIfAbsent(MetricKey.MASTER_LOST_FILE_COUNT.getName(),
         () -> mFileSystemMaster.getLostFiles().size());
   }
 
   @Override
   public void heartbeat() throws InterruptedException {
-    for (long fileId : mFileSystemMaster.getLostFiles()) {
+    Iterator<Long> iter = mBlockMaster.getLostBlocksIterator();
+    while (iter.hasNext()) {
+      long blockId = iter.next();
+      // the file id is the container id of the block id
+      long containerId = BlockId.getContainerId(blockId);
+      long fileId = IdUtils.createFileId(containerId);
       // Throw if interrupted.
       if (Thread.interrupted()) {
         throw new InterruptedException("LostFileDetector interrupted.");
       }
       // update the state
       try (JournalContext journalContext = mFileSystemMaster.createJournalContext();
-          LockedInodePath inodePath =
-              mInodeTree.lockFullInodePath(fileId, LockPattern.WRITE_INODE)) {
+           LockedInodePath inodePath =
+               mInodeTree.lockFullInodePath(fileId, LockPattern.WRITE_INODE)) {
         Inode inode = inodePath.getInode();
         if (inode.getPersistenceState() != PersistenceState.PERSISTED) {
```

Comment:
It's a bit weird here because when we load a file via metadata sync (create an inode), the PersistenceState is not PERSISTED. Then will the fail falsely marked as LOST? WDYT @dbw9580 ?

Reply:
@dbw9580 let's double check this from the metadata sync code